src/phoenix/utilities/ratelimits.py

```diff
@@ -0,0 +1,210 @@
+        enforcement_window_minutes: Optional[int] = None,
+    ) -> None:
+        # default to 1 minute enforcement window
+        enforcement_window_minutes = (
+            enforcement_window_minutes if enforcement_window_minutes is not None else 1
+        )
+        max_tokens = per_minute_rate_limit * enforcement_window_minutes
+        if limits := self._rate_limits[key]:
+            if limit := limits.get(limit_type):
+                limit.refresh_limit(per_minute_rate_limit, max_tokens)
+                return
+        limits[limit_type] = TokenLimiter(
+            per_minute_rate_limit,
+            0,
+            max_tokens,
+        )
+
+    def get_rate_limits(self, key: str) -> Dict[str, TokenLimiter]:
+        return self._rate_limits[key]
+
+    def wait_for_rate_limits(self, key: str, rate_limit_costs: Dict[str, Numeric]) -> None:
+        rate_limits = self._rate_limits[key]
+        for limit_type, cost in rate_limit_costs.items():
+            if limit := rate_limits.get(limit_type):
+                limit.wait_for_then_spend_available_tokens(cost)
+
+    async def async_wait_for_rate_limits(
+        self, key: str, rate_limit_costs: Dict[str, Numeric]
+    ) -> None:
+        rate_limits = self._rate_limits[key]
+        for limit_type, cost in rate_limit_costs.items():
+            if limit := rate_limits.get(limit_type):
+                await limit.async_wait_for_then_spend_available_tokens(cost)
+
+    def spend_rate_limits(self, key: str, rate_limit_costs: Dict[str, Numeric]) -> None:
+        rate_limits = self._rate_limits[key]
+        for limit_type, cost in rate_limit_costs.items():
+            if limit := rate_limits.get(limit_type):
+                limit.spend_tokens(cost)
+
+
+class OpenAIRateLimiter:
+    def __init__(self, api_key: str) -> None:
+        self._store = LimitStore()
+        self._api_key = api_key
+
+    def key(self, model_name: str) -> str:
+        return f"openai:{self._api_key}:{model_name}"
+
+    def set_rate_limits(
+        self, model_name: str, request_rate_limit: Numeric, token_rate_limit: Numeric
+    ) -> None:
+        self._store.set_rate_limit(self.key(model_name), "requests", request_rate_limit)
+        self._store.set_rate_limit(self.key(model_name), "tokens", token_rate_limit)
+
+    def limit(
+        self, model_name: str, token_cost: Numeric
+    ) -> Callable[[Callable[P, T]], Callable[P, T]]:
+        def rate_limit_decorator(fn: Callable[P, T]) -> Callable[P, T]:
+            @wraps(fn)
+            def wrapper(*args: Any, **kwargs: Any) -> T:
+                self._store.wait_for_rate_limits(
+                    self.key(model_name), {"requests": 1, "tokens": token_cost}
+                )
+                result: T = fn(*args, **kwargs)
+                return result
+
+            return wrapper
+
+        return rate_limit_decorator
+
+    def alimit(
+        self,
+        model_name: str,
+        input_cost_fn: Callable[..., Numeric],
```

Comment:
this makes it such that we largely rely on token counters on our end - which in reality is probably a pretty unreliable mechanism. I'd consider getting the costs purely from the APIs if possible.

Reply:
After reading the token counting code I completely agree. I'll change this